HeterogeneousCore/AlpakaInterface/interface/prefixScan.h

```diff
@@ -0,0 +1,228 @@
+        uint32_t const threadDimension(alpaka::getWorkDiv<alpaka::Thread, alpaka::Elems>(acc)[0u]);
+        uint32_t const blockIdx(alpaka::getIdx<alpaka::Grid, alpaka::Blocks>(acc)[0u]);
+
+        auto& ws = alpaka::declareSharedVar<T[warpSize], __COUNTER__>(acc);
+        // first each block does a scan of size warpSize² (better be enough blocks)
+#ifndef NDEBUG
+        [[maybe_unused]] uint32_t const gridDimension(alpaka::getWorkDiv<alpaka::Grid, alpaka::Blocks>(acc)[0u]);
+        ALPAKA_ASSERT_OFFLOAD(gridDimension / threadDimension <= (warpSize * warpSize));
+#endif
+#if 0
+        // this is not yet available in alpaka, see
+        // https://github.com/alpaka-group/alpaka/issues/1648
+        ALPAKA_ASSERT_OFFLOAD(sizeof(T) * gridDimension <= dynamic_smem_size());  // size of psum below
+#endif
+        int off = blockDimension * blockIdx * threadDimension;
+        if (size - off > 0)
+          blockPrefixScan(acc, ci + off, co + off, std::min(int(blockDimension * threadDimension), size - off), ws);
+      }
+    };
+
+    // limited to warpSize⁴ elements
+    template <typename T>
+    struct multiBlockPrefixScanSecondStep {
+      template <typename TAcc>
+      ALPAKA_FN_ACC void operator()(const TAcc& acc, T const* ci, T* co, int32_t size, int32_t numBlocks) const {
+        uint32_t const blockDimension(alpaka::getWorkDiv<alpaka::Block, alpaka::Threads>(acc)[0u]);
+        uint32_t const threadDimension(alpaka::getWorkDiv<alpaka::Thread, alpaka::Elems>(acc)[0u]);
+        uint32_t const threadIdx(alpaka::getIdx<alpaka::Block, alpaka::Threads>(acc)[0u]);
+
+        T* const psum = alpaka::getDynSharedMem<T>(acc);
+
+        // first each block does a scan of size warpSize² (better be enough blocks)
+        ALPAKA_ASSERT_OFFLOAD(static_cast<int32_t>(blockDimension * threadDimension) >= numBlocks);
+        for (int elemId = 0; elemId < static_cast<int>(threadDimension); ++elemId) {
+          int index = +threadIdx * threadDimension + elemId;
+
+          if (index < numBlocks) {
+            int lastElementOfPreviousBlockId = index * blockDimension * threadDimension - 1;
+            psum[index] = (lastElementOfPreviousBlockId < size and lastElementOfPreviousBlockId >= 0)
+                              ? co[lastElementOfPreviousBlockId]
+                              : T(0);
+          }
+        }
+
+        alpaka::syncBlockThreads(acc);
+
+        auto& ws = alpaka::declareSharedVar<T[warpSize], __COUNTER__>(acc);
+        blockPrefixScan(acc, psum, psum, numBlocks, ws);
+
+        for (int elemId = 0; elemId < static_cast<int>(threadDimension); ++elemId) {
+          int first = threadIdx * threadDimension + elemId;
+          for (int i = first + blockDimension * threadDimension; i < size; i += blockDimension * threadDimension) {
+            auto k = i / (blockDimension * threadDimension);
+            co[i] += psum[k];
+          }
+        }
+      }
+    };
+
+  }  // namespace alpakatools
+}  // namespace cms
+
```

Comment:
The `multiBlockPrefixScanSecondStep` kernel is failing because it uses dynamic shared memory, but the amount of memory is not declared:
```suggestion

// declare the amount of block shared memory used by the multiBlockPrefixScanSecondStep kernel
namespace alpaka::trait {

  template <typename TAcc, typename T>
  struct BlockSharedMemDynSizeBytes<cms::alpakatools::multiBlockPrefixScanSecondStep<T>, TAcc> {
    template <typename TVec>
    ALPAKA_FN_HOST_ACC static std::size_t getBlockSharedMemDynSizeBytes(
        cms::alpakatools::multiBlockPrefixScanSecondStep<T> const& /* kernel */,
        TVec const& /* blockThreadExtent */,
        TVec const& /* threadElemExtent */,
        T const* /* ci */,
        T* /* co */,
        int32_t /* size */,
        int32_t numBlocks) {
      return sizeof(T) * numBlocks;
    }
  };

}  // namespace alpaka::trait
```

This seems to fix the unit tests.

Reply:
By the way, this looks like an older version of the multi-block prefix scan, while the CUDA version does not use two separate kernels any more.

Could you add a note that this should be updated ?

Reply:
Sure. Let me know if what I wrote works.

Reply:
Mhm... wrote where ?

Reply:
https://github.com/cms-sw/cmssw/blob/37b2dbac43e608199dc681fc793fe7718cc473cb/HeterogeneousCore/AlpakaInterface/interface/prefixScan.h#L13

(rebasing though)